### PR TITLE
Command Line Interface

### PR DESCRIPTION
**Pull request by burgess01**
# Command Line Interface

## Description

This pull request introduces the main.py file with the basic commands for the software introduced. I also added an __init__ file that will help integrate main.py with the rest of the software. As of creating this pull request, there is no code under either of the sections. This is because we do not have functions from the other team to call yet, and through discussion with the Generate team they asked to handle --force/-f in their own section.

## Linked Issues

closes: #22 

## Type of Change

- [x] Feature
- [ ] Bug fix
- [ ] Documentation

## Contributors

- @burgess01 @watson02 

## Reminder

 - All GitHub Actions should be in a passing state before any pull request is merged.
 - All PRs must be reviewed by at least one team member and one member of the Integration team!
 - Any issues this PR closes are tagged in the description!

### CODE REVIEW

File: gatorgrade/main.py
Comment (Michionlion) [inline gatorgrade/main.py:4]: Could you explain more about why `file` is a global variable here? If it's intended to be a constant, it should probably follow `pylint`'s naming guidelines for constant globals: `FILE` would be the correct name.

File: gatorgrade/main.py
Comment (Michionlion) [inline gatorgrade/main.py:8]: This command setup would require the user to specify a file name (presumably for `gatorgrade.yml`) when running this command; however, I think it would be better if the file location was automatically found, or just assumed to be in a certain relative location.

File: gatorgrade/main.py
Comment (Michionlion) [inline gatorgrade/main.py:13]: Generally it's a good idea to have all options be lower-case -- upper-case options are normally used only when there is already a lower-case version that would conflict. This makes typing the command faster, since the user doesn't need to type a capital letter. There are a fair amount of exceptions in various command line tools, however.

**Review by Michionlion — CHANGES_REQUESTED** [review]

File: gatorgrade/main.py
Comment (burgess01) [reply gatorgrade/main.py]: I intended this to be a global variable since it is something I assumed mutliple commands would probably use, so I thought defining it once and using it from there could be a good idea. I changed the name to be all capitalized to fit better with pylint's naming guidelines!

File: gatorgrade/main.py
Comment (burgess01) [reply gatorgrade/main.py:8]: That makes a lot of sense, thank you! I removed file as a parameter for the gatorgrade command.

File: gatorgrade/main.py
Comment (burgess01) [reply gatorgrade/main.py:13]: Thats true, I made it uppercase as it is what the generate team suggested to me, though if the short version is for ease of access it makes more sense for it to be lowercase. I changed it to be lowercase and I'll follow up with the generate team about the change!

**Review by Michionlion — CHANGES_REQUESTED** [review]
The code itself looks good to me; however as one additional thing I would like to see a basic plumbing test that would use the `CliRunner` testing helper described in Typer's [documentation](https://typer.tiangolo.com/tutorial/testing/), to ensure your Typer configuration is tested. I think that may have to wait until PR #38 is merged, so I would suggest waiting to merge this until you can make that change.

File: tests/test_main.py
Comment (ullrichd21) [inline tests/test_main.py:6]:
```suggestion
def test_gatorgrade_runs_correctly():
```

File: tests/test_main.py
Comment (ullrichd21) [inline tests/test_main.py:16]:
```suggestion
def test_generate_fails_with_existing_yml():
```

**Review by ullrichd21 — CHANGES_REQUESTED** [review]
Left a few minor suggestions. Great work!

**Review by watson02 — COMMENTED** [review]
Good. Ill attended to the comments

**Review by watson02 — COMMENTED** [review]
Looks Good.

File: gatorgrade/main.py
Comment (ullrichd21) [inline gatorgrade/main.py:5]: Just leaving this as a note here, we should probably avoid globals. I think a better way to go about it would be to create a config.yml file that has all your default values you need and then use a singleton config module that you can then reference default/configuration properties from. Just a note for the future, maybe make this an issue/discussion if you all agree?

Thoughts @noorbuchi ?

**Review by ullrichd21 — DISMISSED** [review]
Looks good! Left a comment that you should look at just for future reference but not something that needs to be addressed for this PR!

File: gatorgrade/main.py
Comment (noorbuchi) [reply gatorgrade/main.py:5]: I think this is fine for now, we're going with the assumption that the file will always be called `gatorgrade.yml` and that's fine.

**Review by noorbuchi — APPROVED** [review]

File: gatorgrade/main.py
Comment (noorbuchi) [inline gatorgrade/main.py:11]: can you please make sure that the flag of this argument is specified in the Typer option. Thanks!

**Review by noorbuchi — CHANGES_REQUESTED** [review]

**Review by noorbuchi — APPROVED** [review]

File: gatorgrade/main.py
Comment (Michionlion) [reply gatorgrade/main.py:7]: @ullrichd21 @noorbuchi I'd like to add here that I think GatorGrade should shy away from any configuration that is not in `gatorgrade.yml` or (if absolutely needed) `pyproject.toml`, so that it can work well with the rest of the python ecosystem without adding additional complexity in project structures.

File: .github/cspell.json
Comment (Michionlion) [inline .github/cspell.json:6]: I'm not sure that this change is a good idea. This JSON file is in what is known as "JSONC", or JSON with comments; if the integration team would like to ensure this file is in JSON format, it would be better to remove these comments, as they may have unintended effects if they are added to the JSON structure.

File: gatorgrade/__init__.py
Comment (Michionlion) [inline gatorgrade/__init__.py:1]: This should feature a docstring for the `gatorgrade` package. If the documentation for this package is not ready, I would at least suggest adding a stub like `"""GatorGrade is a program to do X."""`.

File: gatorgrade/main.py
Comment (Michionlion) [inline gatorgrade/main.py:6]: In my understanding, the `gatorgrade.yml` file should be at the root level of the assignment, not inside a `config` directory -- @mariakimheinert, as a potential future user of this project, is this true?

File: gatorgrade/main.py
Comment (Michionlion) [inline gatorgrade/main.py:13]: Instead of referring to the configuration filepath here as `path`, you should chose a more descriptive argument name, such as `--config` or `-c` (or both).

File: gatorgrade/main.py
Comment (Michionlion) [inline gatorgrade/main.py:28]:
```suggestion
    """Generates a gatorgrade.yml file based on the folders and files in the current directory."""
```

File: tests/test_main.py
Comment (Michionlion) [inline tests/test_main.py:6]: Instead of using a word like `correct`/`correctly`, it can be more semantically readable if you use the actual activity it should complete; i.e. this test should be named `test_gatorgrade_runs`.

File: tests/test_main.py
Comment (Michionlion) [inline tests/test_main.py:11]: In the same vein as my comment above, this should be named something like `test_generate_runs` or (if it is indeed checking the contents of the generated yml file) `test_generate_creates_valid_yml`. However, I would suggest that your CLI test suite *only* tests the plumbing; i.e. don't test any of the generated content or behavior, and instead use something like [pytest-mock](https://pypi.org/project/pytest-mock/) to test that the behavior functions written by other teams are called as intended.

File: tests/test_main.py
Comment (Michionlion) [inline tests/test_main.py:16]: This test name works fairly well with the convention outlined above; however `fail` probably should be plural.

File: tests/test_main.py
Comment (Michionlion) [inline tests/test_main.py:22]: `force` is an option, not a command; this may be better named `test_generate_with_force_option_creates_yml`, or in the case that you're mocking the main function written by the generate team, `test_generate_with_force_option_always_calls_generate`.

**Review by Michionlion — CHANGES_REQUESTED** [review]
This looks good! I've given a few suggestions and requests below.

File: gatorgrade/__init__.py
Comment (Michionlion) [inline gatorgrade/__init__.py:1]: My bad on this suggestion... I meant that you should replace `X` with what GatorGrade does at a high-level, which should be clear enough at this point I think.

**Review by Michionlion — CHANGES_REQUESTED** [review]
I think this is mostly good enough to merge, but my one outstanding comment from above about the location of `gatorgrade.yml` still stands and should be addressed before this is merged.

File: gatorgrade/main.py
Comment (mariakimheinert) [reply gatorgrade/main.py:6]: Yes, this is true--`gatorgrade.yml` should be at the root level of the assignment.

Comment (burgess01) [reply gatorgrade/main.py]: Ok! I changed this so that the .yml file will be generated at the root level.

**Review by mariakimheinert — CHANGES_REQUESTED** [review]
Hello @burgess01! It appears that the "Lint Code" task in GitHub Actions is failing. After you address the code linting issues, would you please re-request review from all required reviewers so that everyone knows to look over this PR again? Thank you!

**Comment by burgess01** [comment]
@mariakimheinert Yes! I fixed the rest of the linting errors. The problem with this is that it puts the testing coverage under 60%, as I had to use all of the parameters for the commands in order for them to pass linting checks. Is this okay or should I go back and add tests for these temporary structures? Either way I have re-requested reviews from everyone who had the symbol next to their name!

File: gatorgrade/main.py
Comment (mariakimheinert) [inline gatorgrade/main.py:17]: Marked as a callback, this function will always run for every command, even when the `generate` subcommand is used. This presents a problem since the `gatorgrade` and `gatorgrade generate` commands should do very different things. One way around this is to check the value of the subcommand through the [Typer context](https://typer.tiangolo.com/tutorial/commands/context/#getting-the-context). So, if the value of the invoked subcommand is `None` (which means `generate` was not used), then the behavior of the default `gatorgrade` command should be run. I suggest [adding a script to the `pyproject.toml`](https://python-poetry.org/docs/pyproject/#scripts) to make it easier to run GatorGrade for manually testing the CLI as you make changes. It may be helpful to take a look at [another project's `pyproject.toml`](https://github.com/cmpsc-481-s22-m1/GatorConfig/blob/main/pyproject.toml) for reference.

File: gatorgrade/main.py
Comment (mariakimheinert) [inline gatorgrade/main.py:12]: This format string is not needed--it can just be `FILE`, instead of `f"{FILE}"`.

File: gatorgrade/main.py
Comment (mariakimheinert) [inline gatorgrade/main.py:16]: This line currently throws the error, `Result AttributeError("'PosixPath' object has no attribute 'contains'")`, as `filename` is of type [Path](https://docs.python.org/3/library/pathlib.html), which does not have a `contains` method. Instead, what if you used the [`suffix` attribute](https://docs.python.org/3/library/pathlib.html#pathlib.PurePath.suffix) to check the extension?

File: tests/test_main.py
Comment (mariakimheinert) [inline tests/test_main.py:36]: The expected exit code should probably be 0, unless you are expecting something to go wrong when using the `--force` option.

**Review by mariakimheinert — CHANGES_REQUESTED** [review]
Hello @burgess01, thank you for all your hard work! I have identified some issues with the current implementation that if addressed should result in passing tests and test coverage that meets the minimum requirement. Please let me know if you have any questions about my feedback!

**Review by laurennevill — DISMISSED** [review]
This looks good. Great work

**Review by noorbuchi — DISMISSED** [review]

File: gatorgrade/main.py
Comment (burgess01) [reply gatorgrade/main.py]: @Michionlion I had a quick question about this. So Rian fixed this line, but now it is causing errors in the code. Specifically, now that I fixed this all of the filename declaration information has to be on one line in order to follow formatting rules. This causes line 12 of main.py to be 9 characters over the 79 character limit. I have tried splitting it up onto different lines by splitting it by the different parameters for filename as well as by the parenthesis, but both of these don't follow formatting rules. Do you have any suggestions on how to fix this?

Comment (Michionlion) [reply gatorgrade/main.py]: Run `black` -- if there are conflicting formatting after `black` is run, then the configuration of your tooling should probably be adjusted.

Comment (Michionlion) [reply gatorgrade/main.py]: Okay, I did some investigating and it looks like there's some instructions on what you should do here: https://black.readthedocs.io/en/stable/guides/using_black_with_other_tools.html

You should be trying to keep configuration to just the `pyproject.toml` file as much as possible, so I think this may be a good reason to get rid of `flake8` entirely; pylint does do a lot of what flake8 does, so it wouldn't be a great loss. That would enable you to bypass this issue entirely, since the default configuration for pylint conforms line-length to 88 characters, which is what `black` enforces.

File: pyproject.toml
Comment (Michionlion) [inline pyproject.toml:35]: You don't use the `isort` tool so there is no need to add this configuration.

**Review by Michionlion — CHANGES_REQUESTED** [review]
@watson02 I think there may be a misunderstanding happening here -- I don't think you need to add any extra configuration. Commit 7e3e21abad197eed9d38b9fba6e592b83472e466 is the only commit that you needed to add -- it may actually be simpler to revert the others as it's particularly difficult to review the changes that are spread out among many extremely tiny commits, which isn't really the best practice. Thank you for your earnestness in attempting to resolve these issues, however!

**Comment by watson02** [comment]
@Michionlion ok thank you I will make those changes.

**Review by ullrichd21 — APPROVED** [review]

**Review by Michionlion — CHANGES_REQUESTED** [review]
Hi @burgess01 and @watson02! This PR looks almost good to go! There's just two issues: a bit of unneeded leftover configuration in `pyproject.toml` (annotated below), and an unnecessary `.DS_Store` file accidentally committed. Once those are fixed re-request my review and I think you'll be good to merge!

**Review by noorbuchi — DISMISSED** [review]
Looks great!

**Review by laurennevill — DISMISSED** [review]

**Review by noorbuchi — APPROVED** [review]

**Review by laurennevill — APPROVED** [review]

**Review by ullrichd21 — APPROVED** [review]

**Review by Michionlion — APPROVED** [review]
Thank you for all of your hard work on these foundational changes, @burgess01 and @watson02! I think this is finally ready to merge!

**Pull request merged**
